// infrastructure/mailer/MailSender.ts
/**
 * Created by x on 03/09/16.
 */
const nodemailer = require("nodemailer");
import Configuration from "Configuration";
import {MailOptions} from "framework/MailOptions";
const SMTP_CONFIG = Configuration.emailConfig.smtpConfig;

export interface MailSender {
    sendMail(options?: MailOptions): void;
}

export class MailSenderImpl implements MailSender {
    private transporter: any;
    private nodemailer: any;

    constructor(_nodemailer: any, transporter: any) {
        this.nodemailer = _nodemailer;
        this.transporter = transporter || this.nodemailer.createTransport(SMTP_CONFIG);
    }

    public sendMail(options?: MailOptions): void {
        options = options || {};
        const mailOpt = {
            from: "\"eHealthScanner\" <hello@ehealthscanner.com>",
            to: options.to || "ehealthscannertest1@mailinator.com",
            subject: options.subject || "ehealth mailer",
            text: options.text || "eHealthScanner!!",
            html: options.html || null,
            attachments: options.attachments || [],
        };

        console.error(">>>>>NODEMAILER CONFIGURATION............ ");
        console.error(">smtpConfiguration");
        console.error(SMTP_CONFIG);
        console.error(">mailOptions");
        console.error(mailOpt);
        try {
            this.transporter.sendMail(mailOpt);
        } catch (e) {
            console.log(e);
        }
    }
}

export interface SmtpConfiguration {
    host?: string;
    port?: string;
    secure?: boolean; // true=SSL
    auth: EmailAccount;
}

export interface EmailAccount {
    user: string;
    pass: string;
}

export function MakeMailSender(_nodemailer?: any, transporter?: any): MailSender {
    return new MailSenderImpl(_nodemailer || nodemailer, transporter);
}


// function MakeSmtpConfiguration(smtpconf: smtpConfiguration) {
//     return new smtpConfigurationCreator(smtpconf).create();
// }

// class smtpConfigurationCreator {
//     private host: string;
//     private port: string;
//     private secure: boolean;
//     private auth: emailAccount;
//
//     constructor(conf: smtpConfiguration) {
//         this.host = conf.host || "smtp.gmail.com";
//         this.port = conf.port || "465";
//         this.secure = conf.secure || true;
//         this.auth.user = conf.auth.user || "xxxxx@apiumtech.com";
//         this.auth.pass = conf.auth.pass || "some_password";
//     }
//
//     public create() {
//         return {
//             host: this.host,
//             port: this.port,
//             secure: this.secure,
//             auth: {
//                 user: this.auth.user,
//                 pass: this.auth.pass
//             },
//         };
//     }
// }
