// domain/plan-case/PlanCaseReadModel.ts
/**
 * Created by x on 30/08/16.
 */


import {InvalidArgumentException} from "domain/Exception";
import {Entity, AbstractEntity, EntityDbRow, EntityDTO} from "domain/Entity";
import {Instance} from "sequelize";
import {MailOptions} from "framework/MailOptions";
import validator from "framework/Validator";
import {MakeAttachmentService, Attachment} from "infrastructure/AttachmentService";
import {ConverterTool} from "framework/CurrencyConverterTool";
import Configuration from "Configuration";
import EmailTemplate from "domain/plan-case/EmailTemplate";


export interface Person {
    id?: string;
    age: string;
    gender?: string;
    nationality?: any;
    relationShip?: string;
    masterAreaOfCover?: any;
}

export interface UserNotificationParams {
    planId: string;
    numberInsured: string;
    countryOfResidence: string;
    areaOfCover: string;
    excess: string;
    totalAnnualPremium: string;
    persons: Person[];
    userInfo: UserInfoParams;
    refId: string;
}

export interface UserInfoParams {
    firstName: string;
    lastName: string;
    email: string;
    phone?: string;
    comment?: string;
    location?: string;
    preferred?: string;
    request?: boolean;
    questions?: boolean;
}

export interface PlanCaseReadModel extends Entity<PlanCaseReadModelActiveRecord, PlanCaseReadModelDTO> {
    getEmailContentForUser(data: UserNotificationParams, page: string): MailOptions;
    getEmailContentForAdmin(data: UserNotificationParams, page: string): MailOptions;
    convertPrice(converter: ConverterTool): void;
}

export interface PlanCaseReadModelDTO extends EntityDTO {
    uuid: string;
    plan_case_id: string;
    plan_id: string;
    plan_title: string;
    plan_name: string;
    company_name: string;
    area_of_cover: string;
    master_areas_of_cover: string;
    currency: string;
    excess: string;
    maximum_cover: number;
    maximum_cover_periodicity: string;
    excess_periodicity: string;
    outpatient_cover_maximum: string;
    outpatient_cover_periodicity: string;
    outpatient_cover: number;
    outpatient_cover_information: string;
    dental_cover_maximum: string;
    dental_cover_periodicity: string;
    dental_cover: number;
    dental_cover_information: string;
    maternity_cover_maximum: string;
    maternity_cover_periodicity: string;
    maternity_cover: number;
    maternity_cover_information: string;
    optical_cover_maximum: string;
    optical_cover_periodicity: string;
    optical_cover: number;
    optical_cover_information: string;
    other_highlights: string;
    logo_url: string;
    benefits_schedule_pdf_url: string;
    general_conditions_pdf_url: string;
    total_annual_premium: string;
    participants: number;
}

export class InvalidWorkspaceArgumentException extends InvalidArgumentException {
}

export function PlanCaseReadModelFromActiveRecord(record: PlanCaseReadModelActiveRecord): PlanCaseReadModel {
    return new PlanCaseReadModelImpl(record);
}

export interface PlanCaseReadModelDbRow extends EntityDbRow {
    uuid: string;
    plan_id: string;
    plan_case_id: string;
    plan_title: string;
    plan_name: string;
    company_name: string;
    area_of_cover: string;
    master_areas_of_cover: string;
    currency: string;
    excess: string;
    maximum_cover: number;
    maximum_cover_periodicity: string;
    excess_periodicity: string;
    outpatient_cover_maximum: string;
    outpatient_cover_periodicity: string;
    outpatient_cover: number;
    outpatient_cover_information: string;
    dental_cover_maximum: string;
    dental_cover_periodicity: string;
    dental_cover: number;
    dental_cover_information: string;
    maternity_cover_maximum: string;
    maternity_cover_periodicity: string;
    maternity_cover: number;
    maternity_cover_information: string;
    optical_cover_maximum: string;
    optical_cover_periodicity: string;
    optical_cover: number;
    optical_cover_information: string;
    other_highlights: string;
    logo_url: string;
    benefits_schedule_pdf_url: string;
    general_conditions_pdf_url: string;
    application_form_pdf_url: string;
    co_branding: string;
    co_branding_link: string;
    electronic_signature_application_link: string;
    total_annual_premium: string;
    participants: number;
}


export interface PlanCaseReadModelActiveRecord extends Instance<PlanCaseReadModelDbRow>, PlanCaseReadModelDbRow {

}

class PlanCaseReadModelImpl extends AbstractEntity<PlanCaseReadModelActiveRecord, PlanCaseReadModelDTO> implements PlanCaseReadModel {

    public convertPrice(converter: ConverterTool) {
         if (this.record.currency === "EUR") {
             const converted = converter.convertEurToUSD({
                 currency: "EUR",
                 quantity: this.record.total_annual_premium
             });

             this.record.total_annual_premium = converted.quantity;
         }
    }

    protected _toJSON(): PlanCaseReadModelDTO {
        return {
            uuid: this.record.uuid,
            plan_case_id: this.record.plan_case_id,
            plan_id: this.record.plan_id,
            plan_title: this.record.plan_title,
            plan_name: this.record.plan_name,
            company_name: this.record.company_name,
            area_of_cover: this.record.area_of_cover,
            master_areas_of_cover: this.record.master_areas_of_cover,
            currency: this.record.currency,
            excess: this.record.excess,
            maximum_cover: this.record.maximum_cover,
            maximum_cover_periodicity: this.record.maximum_cover_periodicity,
            excess_periodicity: this.record.excess_periodicity,
            outpatient_cover_maximum: this.record.outpatient_cover_maximum,
            outpatient_cover_periodicity: this.record.outpatient_cover_periodicity,
            outpatient_cover: this.record.outpatient_cover,
            outpatient_cover_information: this.record.outpatient_cover_information,
            dental_cover_maximum: this.record.dental_cover_maximum,
            dental_cover_periodicity: this.record.dental_cover_periodicity,
            dental_cover: this.record.dental_cover,
            dental_cover_information: this.record.dental_cover_information,
            maternity_cover_maximum: this.record.maternity_cover_maximum,
            maternity_cover_periodicity: this.record.maternity_cover_periodicity,
            maternity_cover: this.record.maternity_cover,
            maternity_cover_information: this.record.maternity_cover_information,
            optical_cover_maximum: this.record.optical_cover_maximum,
            optical_cover_periodicity: this.record.optical_cover_periodicity,
            optical_cover: this.record.optical_cover,
            optical_cover_information: this.record.optical_cover_information,
            other_highlights: this.record.other_highlights,
            logo_url: this.record.logo_url,
            benefits_schedule_pdf_url: this.record.benefits_schedule_pdf_url,
            general_conditions_pdf_url: this.record.general_conditions_pdf_url,
            total_annual_premium: this.record.total_annual_premium,
            participants: this.record.participants || 1,
        };
    }

    public getEmailContentForUser(data: UserNotificationParams, page: string): MailOptions {
        return page === "askUs" ? this._getEmailContentForUserAskUs(data) : (this.record.co_branding ?
            this._getEmailContentForUserCoBranding(data) :
            this._getEmailContentForUserNoCoBranding(data));
    }

    public getEmailContentForAdmin(data: UserNotificationParams, page: string): MailOptions {
        return {
            subject: (page === "askUs" ? "Ask Us" : "New") + " " + (this.record.co_branding ? "co-branding" : "(no co-branding)") + " purchase request for " + this.record.company_name + " " + this.record.plan_name + ", " + data.userInfo.lastName,
            text: "Hi!,\n\n" +

            "This user is interested in the following plan:\n\n" +
            "AFFILIATE CODE: " + data.refId + "\n\n" +
            "USER\n" +
            "-------------- \n" +
            "- FIRST AND LAST NAME: " + data.userInfo.firstName + " " + data.userInfo.lastName + "\n" +
            "- EMAIL ADDRESS: "       + data.userInfo.email   + "\n" +
            "- PHONE NUMBER: "        + (validator.isNotUndefined(data.userInfo.phone) ? data.userInfo.phone : "") + " (optional)\n" +
            (page === "askUs" ? (
                "- REQUEST A PHONE CALLBACK: "           + (validator.isNotUndefined(data.userInfo.request)   ? data.userInfo.request     : "")   + "\n" +
                "- PREFERRED CALLBACK DATE AND TIME: "  + (validator.isNotUndefined(data.userInfo.preferred) ? data.userInfo.preferred   : "")   + "\n" +
                "- CALLBACK LOCATION (City, Country): "  + (validator.isNotUndefined(data.userInfo.location)  ? data.userInfo.location    : "")   + "\n" +
                "- QUESTIONS: "                          + (validator.isNotUndefined(data.userInfo.questions) ? data.userInfo.questions   : "")   + "\n\n"
            ) : "\n") +

            "REQUEST\n" +
            "-------------- \n" +
            "- NUMBER OF INDIVIDUALS: "       + data.numberInsured + "\n" +
            "- COUNTRY OF RESIDENCE: "        + data.countryOfResidence + "\n" +
            "- COVER REQUIREMENTS: "          + (this.record.outpatient_cover === 1 ? "Outpatient cover " : "")
                                            + (this.record.dental_cover === 1 ? "Dental cover " : "")
                                            + (this.record.maternity_cover === 1 ? "Maternity cover " : "")
                                            + (this.record.optical_cover === 1 ? "Optical cover" : "") + "\n" +
            "- MAIN INSURED AGE: "            + data.persons[0].age + "\n" +
            "- MAIN INSURED GENDER: "         + data.persons[0].gender + "\n" +
            "- MAIN INSURED NATIONALITY: "    + data.persons[0].nationality.country_name + "\n" +
            "- IDEAL AREA OF COVER: "         + data.persons[0].masterAreaOfCover.master_area_name + "\n" +

            (data.persons.length > 1 ? this._getPersonsInfo(data.persons) + "\n" : "") +
            "- COMMENTS: " + (validator.isNotUndefined(data.userInfo.comment) ? data.userInfo.comment : "") + "\n\n" +


            "RESULT\n" +
            "-------------- \n" +
            "- COMPANY NAME: "          + this.record.company_name  + "\n" +
            "- PLAN NAME: "             + this.record.plan_name     + "\n" +
            "- PLAN ID: "               + this.record.plan_id       + "\n" +
            "- PLAN CASES: "            + this.record.plan_case_id  + "\n" +
            "- AREA OF COVER: "         + this.record.area_of_cover + "\n" +
            "- MAXIMUM COVER: "         + this.record.currency      + " "
                                        + this.record.maximum_cover + " "
                                        + this.record.maximum_cover_periodicity         + "\n" +
            "- EXCESS: "                + this.record.currency                          + " "
                                        + this.record.excess                            + " "
                                        + this.record.excess_periodicity                + "\n" +
            "- OUTPATIENT COVER: "      + (this.record.outpatient_cover === 0 ? "NO" : (
                                          this.record.currency                      + " "
                                        + this.record.optical_cover_maximum         + " "
                                        + this.record.outpatient_cover_periodicity  + " "
                                        + this.record.outpatient_cover_information).trim()) + "\n" +
            "- DENTAL COVER: "          + (this.record.dental_cover === 0 ? "NO" : (
                                          this.record.currency                      + " "
                                        + this.record.dental_cover_maximum          + " "
                                        + this.record.dental_cover_periodicity      + " "
                                        + this.record.dental_cover_information).trim())     + "\n" +
            "- MATERNITY COVER: "       + (this.record.maternity_cover === 0 ? "NO" : (
                                          this.record.currency                      + " "
                                        + this.record.maternity_cover_maximum       + " "
                                        + this.record.maternity_cover_periodicity   + " "
                                        + this.record.maternity_cover_information).trim())  + "\n" +
            "- OPTICAL COVER: "         + (this.record.optical_cover === 0 ? "NO" : (
                                          this.record.currency                      + " "
                                        + this.record.optical_cover_maximum         + " "
                                        + this.record.optical_cover_periodicity     + " "
                                        + this.record.optical_cover_information).trim())    + "\n" +
            "- OTHER HIGHLIGHTS: "        + this.record.other_highlights                  + "\n" +
            "- ESTIMATED MONTHLY COST: "  + "USD " + this.record.total_annual_premium
        };
    }

    private _getEmailContentForUserCoBranding(data: UserNotificationParams): MailOptions {
        return {
            attachments: [
                this._getattachmentsPDF(`${Configuration.ROOT_PATH}upload/documents/` + this.record.general_conditions_pdf_url, this.record.general_conditions_pdf_url),
                this._getattachmentsPDF(`${Configuration.ROOT_PATH}upload/documents/` + this.record.benefits_schedule_pdf_url, this.record.benefits_schedule_pdf_url)
            ],
            subject: "Thank you for your inquiry",
            html: ` 
                    <!doctype html>
                    <html>
                      <head>
                        <meta name="viewport" content="width=device-width" />
                        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
                        <title>Simple Transactional Email</title>
                        <style>` + EmailTemplate.CSS + `</style>
                      </head>
                      <body class="">
                        <table border="0" cellpadding="0" cellspacing="0" class="body">
                          <tr>
                            <td>&nbsp;</td>
                            <td class="container">
                              <p><img src="https://ehealthscanner.com/assets/img/eHealthScanner_.png" alt="ehealthscanner" title="ehealthscanner" ></p>
                              <div class="content">
                    
                                <!-- START CENTERED WHITE CONTAINER -->
                                <span class="preheader">ehealthscanner. Thank you for your inquiry.</span>
                                <table class="main">
                    
                                  <!-- START MAIN CONTENT AREA -->
                                  <tr>
                                    <td class="wrapper">
                                      <table border="0" cellpadding="0" cellspacing="0">
                                        <tr>
                                          <td>
                                              <p>Dear ` + data.userInfo.firstName + " " + data.userInfo.lastName + `,</p>
                                            
                                            <p>Thank you for using ehealthscanner! Attached to this email is your insurance plan's brochure and general conditions.</p>
                    
                                            <p>Please review your chosen plan's information and complete your application below:</p>
                    
                                            <hr>
                    
                                            <p><strong>Number of individuals: </strong>&nbsp; <span>` +  data.numberInsured + `</span></p>
                    
                                            <p><strong>Nationality of the main insured: </strong>&nbsp; 
                                                <span>` +  data.persons[0].nationality.country_name + `</span>
                                            </p>
                    
                                            <p><strong>Country where you will reside when your cover will take effect: </strong>&nbsp; <span>` + data.countryOfResidence + `</span></p>
                    
                                            <p><strong>Company: </strong> &nbsp; <span>` + this.record.company_name + `</span></p>
                    
                                            <p><strong>Plan name: </strong> &nbsp; <span>` + this.record.plan_name + `</span></p>
                    
                                            <p><strong>Area of cover: </strong>&nbsp; <span >` + this.record.area_of_cover + `</span></p>
                    
                                            <p><strong>Maximum cover: </strong>&nbsp; <span>` + this.record.currency + `&nbsp;` + this.record.maximum_cover + `&nbsp;` + this.record.maximum_cover_periodicity + `</span></p>
                    
                                            <p><strong>Excess: </strong>&nbsp; <span >` + this.record.currency + `&nbsp;` + this.record.excess + `&nbsp;` + this.record.excess_periodicity + `</span></p>
                    
                                            <p><strong>Outpatient cover: </strong>&nbsp; <span >` + (this.record.outpatient_cover === 0 ? "NO" :
                                            (this.record.currency + `&nbsp;` + this.record.outpatient_cover_maximum + `&nbsp;` + this.record.outpatient_cover_periodicity    + `&nbsp;` + this.record.outpatient_cover_information).trim()) + `</span></p>
                    
                                            <p><strong>Dental cover: </strong>&nbsp; <span >` + (this.record.dental_cover === 0 ? "NO" : (this.record.currency + `&nbsp;` +
                                            this.record.dental_cover_maximum + `&nbsp;` + this.record.dental_cover_periodicity + `&nbsp;` + this.record.dental_cover_information).trim()) + `</span></p>
                    
                                            <p><strong>Maternity cover: </strong>&nbsp; <span >` + (this.record.maternity_cover === 0 ? "NO" : (this.record.currency + `&nbsp;` +
                                            this.record.maternity_cover_maximum + `&nbsp;` + this.record.maximum_cover_periodicity + `&nbsp;` + this.record.maternity_cover_information).trim()) + `</span></p>
                    
                                            <p><strong>Optical cover: </strong>&nbsp; <span >` + (this.record.optical_cover === 0 ? "NO" : ( this.record.currency + `&nbsp;` +
                                            this.record.optical_cover_maximum + `&nbsp;` + this.record.optical_cover_periodicity + `&nbsp;` + this.record.optical_cover_information).trim()) + `</span></p>
                    
                                            <p><strong>Other highlights: </strong>&nbsp; <span >` + this.record.other_highlights + `</span></p>
                    
                                            <p><strong>Estimated monthly cost (based on the annual premium): &nbsp; <span > USD ` + this.record.total_annual_premium + `</span></strong></p>
                    
                                            <hr>
                    
                                          <p>What's next?</p>
                    
                                          <p>Let's start your application. Click on the button below to apply directly on the insurance company's website:</p>
                    
                                          <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                              <tbody>
                                                <tr>
                                                  <td align="center">
                                                    <table border="0" cellpadding="0" cellspacing="0">
                                                      <tbody>
                                                        <tr>
                                                          <td> <a href="` + this.record.co_branding_link + `" target="_blank">Apply Now</a> </td>
                                                        </tr>
                                                      </tbody>
                                                    </table>
                                                  </td>
                                                </tr>
                                              </tbody>
                                            </table>
                    
                                          <hr>
                                          <p>If the insurance company requires that you scan documents and you don't own a scanner, you can download the below application to scan your application on your mobile device!</p>
                    
                                            <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                              <tbody>
                                                <tr>
                                                  <td align="center">
                                                    <table border="0" cellpadding="0" cellspacing="0">
                                                      <tbody>
                                                        <tr>
                                                          <td> <a href="https://play.google.com/store/apps/details?id=com.intsig.camscanner&hl=en" target="_blank">CamScanner Android</a> </td>
                                                        </tr>
                                                      </tbody>
                                                    </table>
                                                  </td>
                                                </tr>
                                              </tbody>
                                            </table>
                    
                                            <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                              <tbody>
                                                <tr>
                                                  <td align="center">
                                                    <table border="0" cellpadding="0" cellspacing="0">
                                                      <tbody>
                                                        <tr>
                                                          <td> <a href="https://itunes.apple.com/us/app/camscanner-free-pdf-document/id388627783?mt=8" target="_blank">CamScanner iOS</a> </td>
                                                        </tr>
                                                      </tbody>
                                                    </table>
                                                  </td>
                                                </tr>
                                              </tbody>
                                            </table>
                    
                                            <hr>
                                            <p>We will insure the application is swiftly processed and <company> will soon get back to you with their response. </p>
                    
                                            <p>Below is the link to create your ehealthscanner customer portal access (please click on "sign up" on the top-right of the page). Through your customer portal, 
                                            you can access useful information in the "help center" section start a ticket and upload documents securely if you have questions or a claim you would need help with!</p>
                                          
                                            <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                              <tbody>
                                                <tr>
                                                  <td align="center">
                                                    <table border="0" cellpadding="0" cellspacing="0">
                                                      <tbody>
                                                        <tr>
                                                          <td> <a href="https://support.ehealthscanner.com/portal/home" target="_blank">Customer Portal</a> </td>
                                                        </tr>
                                                      </tbody>
                                                    </table>
                                                  </td>
                                                </tr>
                                              </tbody>
                                            </table>
                    
                                            <p>If you have any questions or would like help filling up your application please feel free to ask by return of email or live on our website's chat!</p>
                    
                                            <p>Kind regards,</p>
                                            <p>The ehealthscanner customer support team</p>
                                          </td>
                                        </tr>
                                      </table>
                                    </td>
                                  </tr>
                    
                                  <!-- END MAIN CONTENT AREA -->
                                  </table>`
            + EmailTemplate.FOOTER +
            `<!-- END CENTERED WHITE CONTAINER --></div>
                            </td>
                            <td>&nbsp;</td>
                          </tr>
                        </table>
                      </body>
                    </html>
                `
        };
    }

    private _getEmailContentForUserNoCoBranding(data: any): MailOptions {
        return {
            attachments: [
                this._getattachmentsPDF(`${Configuration.ROOT_PATH}upload/documents/` + this.record.application_form_pdf_url, this.record.application_form_pdf_url),
                this._getattachmentsPDF(`${Configuration.ROOT_PATH}upload/documents/` + this.record.general_conditions_pdf_url, this.record.general_conditions_pdf_url),
                this._getattachmentsPDF(`${Configuration.ROOT_PATH}upload/documents/` + this.record.benefits_schedule_pdf_url, this.record.benefits_schedule_pdf_url)
            ],
            subject: "Thank you for your inquiry",
            html: ` 
                <!doctype html>
                <html>
                  <head>
                    <meta name="viewport" content="width=device-width" />
                    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
                    <title>Simple Transactional Email</title>
                    <style>` + EmailTemplate.CSS + `</style>
                  </head>
                  <body class="">
                    <table border="0" cellpadding="0" cellspacing="0" class="body">
                      <tr>
                        <td>&nbsp;</td>
                        <td class="container">
                          <p><img src="https://ehealthscanner.com/assets/img/eHealthScanner_.png" alt="ehealthscanner" title="ehealthscanner" ></p>
                          <div class="content">
                
                            <!-- START CENTERED WHITE CONTAINER -->
                            <span class="preheader">ehealthscanner. Thank you for your inquiry.</span>
                            <table class="main">
                
                              <!-- START MAIN CONTENT AREA -->
                              <tr>
                                <td class="wrapper">
                                  <table border="0" cellpadding="0" cellspacing="0">
                                    <tr>
                                      <td>
                                          <p>Dear ` + data.userInfo.name + `,</p>
                                        
                                        <p>Thank you for using ehealthscanner! Attached to this email is your insurance plan's brochure
                                         and general conditions.</p>
                
                                        <p>Please review your chosen plan's information and complete your application below:</p>
                
                                        <hr>
                                        <p><strong>Number of individuals: </strong>&nbsp; <span>` + data.numberInsured + `</span></p>
                    
                                        <p><strong>Nationality of the main insured: </strong>&nbsp; 
                                            <span>` +  data.persons[0].nationality.country_name + `</span>
                                        </p>
                
                                        <p><strong>Country where you will reside when your cover will take effect: </strong>&nbsp; <span>` + data.countryOfResidence + `</span></p>
                
                                        <p><strong>Company: </strong> &nbsp; <span>` + this.record.company_name + `</span></p>
                
                                        <p><strong>Plan name: </strong> &nbsp; <span>` + this.record.plan_name + `</span></p>
                
                                        <p><strong>Area of cover: </strong>&nbsp; <span >` + this.record.area_of_cover + `</span></p>
                
                                        <p><strong>Maximum cover: </strong>&nbsp; <span>` + this.record.currency + `&nbsp;` + this.record.maximum_cover + `&nbsp;` + this.record.maximum_cover_periodicity + `</span></p>
                
                                        <p><strong>Excess: </strong>&nbsp; <span >` + this.record.currency + `&nbsp;` + this.record.excess + `&nbsp;` + this.record.excess_periodicity + `</span></p>
                
                                        <p><strong>Outpatient cover: </strong>&nbsp; <span >` + (this.record.outpatient_cover === 0 ? "NO" :
                                        (this.record.currency + `&nbsp;` + this.record.outpatient_cover_maximum + `&nbsp;` + this.record.outpatient_cover_periodicity    + `&nbsp;` + this.record.outpatient_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Dental cover: </strong>&nbsp; <span >` + (this.record.dental_cover === 0 ? "NO" : (this.record.currency + `&nbsp;` +
                                        this.record.dental_cover_maximum + `&nbsp;` + this.record.dental_cover_periodicity + `&nbsp;` + this.record.dental_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Maternity cover: </strong>&nbsp; <span >` + (this.record.maternity_cover === 0 ? "NO" : (this.record.currency + `&nbsp;` +
                                        this.record.maternity_cover_maximum + `&nbsp;` + this.record.maximum_cover_periodicity + `&nbsp;` + this.record.maternity_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Optical cover: </strong>&nbsp; <span >` + (this.record.optical_cover === 0 ? "NO" : ( this.record.currency + `&nbsp;` +
                                        this.record.optical_cover_maximum + `&nbsp;` + this.record.optical_cover_periodicity + `&nbsp;` + this.record.optical_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Other highlights: </strong>&nbsp; <span >` + this.record.other_highlights + `</span></p>
                
                                        <p><strong>Estimated monthly cost (based on the annual premium): &nbsp; <span > USD ` + this.record.total_annual_premium + `</span></strong></p>
                
                                        <hr>
                
                                      <p>What's next?</p>
                
                                      <p>Let's start your application, here are the possible options:</p>
                
                                      <p>Fill and sign your application electronically (for the best possible electronic signing experience, please use a large touch screen).</p>
                
                                      <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                          <tbody>
                                            <tr>
                                              <td align="center">
                                                <table border="0" cellpadding="0" cellspacing="0">
                                                  <tbody>
                                                    <tr>
                                                      <td> <a href=` + this.record.electronic_signature_application_link + ` target="_blank">Apply electronically</a></td>
                                                    </tr>
                                                  </tbody>
                                                </table>
                                              </td>
                                            </tr>
                                          </tbody>
                                        </table>
                                         
                                      <p>Or you can print the attached application form to fill, sign, scan and upload it through your customer portal
                                      (see below to create your access). You could
                                      also return it to us by email at hello@ehealthscanner.com (please encrypt the email if you decide to do so).</p>
                
                                      <hr>
                
                                      <p>If you don't own a scanner, you can download the below application to scan your application on your mobile device!</p>
                
                                        <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                          <tbody>
                                            <tr>
                                              <td align="center">
                                                <table border="0" cellpadding="0" cellspacing="0">
                                                  <tbody>
                                                    <tr>
                                                      <td> <a href="https://play.google.com/store/apps/details?id=com.intsig.camscanner&hl=en" target="_blank">CamScanner Android</a> </td>
                                                    </tr>
                                                  </tbody>
                                                </table>
                                              </td>
                                            </tr>
                                          </tbody>
                                        </table>
                
                                        <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                          <tbody>
                                            <tr>
                                              <td align="center">
                                                <table border="0" cellpadding="0" cellspacing="0">
                                                  <tbody>
                                                    <tr>
                                                      <td> <a href="https://itunes.apple.com/us/app/camscanner-free-pdf-document/id388627783?mt=8" target="_blank">CamScanner iOS</a> </td>
                                                    </tr>
                                                  </tbody>
                                                </table>
                                              </td>
                                            </tr>
                                          </tbody>
                                        </table>
                
                                        <hr>
                                        <p>We will insure the application is swiftly processed and <company> will soon get back to you with their response. </p>
                
                                        <p>Below is the link to create your ehealthscanner customer portal access (please click on "sign up" on the top-right of the page). Through your customer portal, 
                                        you can access useful information in the "help center" section start a ticket and upload documents securely if you have questions or a claim you would need help with!</p>
                                      
                                        <table border="0" cellpadding="0" cellspacing="0" class="btn btn-primary">
                                          <tbody>
                                            <tr>
                                              <td align="center">
                                                <table border="0" cellpadding="0" cellspacing="0">
                                                  <tbody>
                                                    <tr>
                                                      <td> <a href="https://support.ehealthscanner.com/portal/home" target="_blank">Customer Portal</a> </td>
                                                    </tr>
                                                  </tbody>
                                                </table>
                                              </td>
                                            </tr>
                                          </tbody>
                                        </table>
                
                                        <p>If you have any questions or would like help filling up your application please feel free to ask by return of email or live on our website's chat!</p>
                
                                        <p>Kind regards,</p>
                                        <p>The ehealthscanner customer support team</p>
                                      </td>
                                    </tr>
                                  </table>
                                </td>
                              </tr>
                
                              <!-- END MAIN CONTENT AREA -->
                              </table>`
            + EmailTemplate.FOOTER +
            `<!-- END CENTERED WHITE CONTAINER --></div>
                        </td>
                        <td>&nbsp;</td>
                      </tr>
                    </table>
                  </body>
                </html>
                `
        };
    }

    private _getEmailContentForUserAskUs(data: any): MailOptions {
        return {
            attachments: [
                this._getattachmentsPDF(`${Configuration.ROOT_PATH}upload/documents/` + this.record.general_conditions_pdf_url, this.record.general_conditions_pdf_url),
                this._getattachmentsPDF(`${Configuration.ROOT_PATH}upload/documents/` + this.record.benefits_schedule_pdf_url, this.record.benefits_schedule_pdf_url)
            ],
            subject: "Your request for information - [" + this.record.company_name + "] - [" + data.userInfo.name + "]",
            html: ` 
                <!doctype html>
                <html>
                  <head>
                    <meta name="viewport" content="width=device-width" />
                    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
                    <title>Simple Transactional Email</title>
                    <style>` + EmailTemplate.CSS + `</style>
                  </head>
                  <body class="">
                    <table border="0" cellpadding="0" cellspacing="0" class="body">
                      <tr>
                        <td>&nbsp;</td>
                        <td class="container">
                          <p><img src="https://ehealthscanner.com/assets/img/eHealthScanner_.png" alt="ehealthscanner" title="ehealthscanner" ></p>
                          <div class="content">
                
                            <!-- START CENTERED WHITE CONTAINER -->
                            <span class="preheader">ehealthscanner. Thank you for your inquiry.</span>
                            <table class="main">
                
                              <!-- START MAIN CONTENT AREA -->
                              <tr>
                                <td class="wrapper">
                                  <table border="0" cellpadding="0" cellspacing="0">
                                    <tr>
                                      <td>
                                          <p>Dear ` + data.userInfo.name + `,</p>
                                        
                                        <p>Thank you for using ehealthscanner! Attached to this email is the brochure and general conditions of the plan you expressed an interest in.</p>
                
                                        <p>Also, here is a recall of your search as well as the results:</p>
                
                                        <hr>
                                        <p><strong>Number of individuals: </strong>&nbsp; <span>` + data.numberInsured + `</span></p>
                    
                                        <p><strong>Nationality of the main insured: </strong>&nbsp; 
                                            <span>` +  data.persons[0].nationality.country_name + `</span>
                                        </p>
                
                                        <p><strong>Country where you will reside when your cover will take effect: </strong>&nbsp; <span>` + data.countryOfResidence + `</span></p>
                
                                        <p><strong>Insurance company name: </strong> &nbsp; <span>` + this.record.company_name + `</span></p>
                
                                        <p><strong>Chosen plan: </strong> &nbsp; <span>` + this.record.plan_name + `</span></p>
                
                                        <p><strong>Area of cover: </strong>&nbsp; <span >` + this.record.area_of_cover + `</span></p>
                
                                        <p><strong>Maximum cover: </strong>&nbsp; <span>` + this.record.currency + `&nbsp;` + this.record.maximum_cover + `&nbsp;` + this.record.maximum_cover_periodicity + `</span></p>
                
                                        <p><strong>Excess: </strong>&nbsp; <span >` + this.record.currency + `&nbsp;` + this.record.excess + `&nbsp;` + this.record.excess_periodicity + `</span></p>
                
                                        <p><strong>Outpatient cover: </strong>&nbsp; <span >` + (this.record.outpatient_cover === 0 ? "NO" :
                                        (this.record.currency + `&nbsp;` + this.record.outpatient_cover_maximum + `&nbsp;` + this.record.outpatient_cover_periodicity    + `&nbsp;` + this.record.outpatient_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Dental cover: </strong>&nbsp; <span >` + (this.record.dental_cover === 0 ? "NO" : (this.record.currency + `&nbsp;` +
                                        this.record.dental_cover_maximum + `&nbsp;` + this.record.dental_cover_periodicity + `&nbsp;` + this.record.dental_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Maternity cover: </strong>&nbsp; <span >` + (this.record.maternity_cover === 0 ? "NO" : (this.record.currency + `&nbsp;` +
                                        this.record.maternity_cover_maximum + `&nbsp;` + this.record.maximum_cover_periodicity + `&nbsp;` + this.record.maternity_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Optical cover: </strong>&nbsp; <span >` + (this.record.optical_cover === 0 ? "NO" : ( this.record.currency + `&nbsp;` +
                                        this.record.optical_cover_maximum + `&nbsp;` + this.record.optical_cover_periodicity + `&nbsp;` + this.record.optical_cover_information).trim()) + `</span></p>
                
                                        <p><strong>Other highlights: </strong>&nbsp; <span >` + this.record.other_highlights + `</span></p>
                
                                        <p><strong>Estimated monthly cost (based on the annual premium): &nbsp; <span > USD ` + this.record.total_annual_premium + `</span></strong></p>
                
                                        <hr>
                
                                        <p>Please ask us the questions that you have by return of email or we can set up a live chat session during office hours, please let us know your availabilities:</p>
                                        
                                        <p>Monday to Friday 9 a.m. to 6 p.m. (GMT +7)</p>
                                        
                                        <p>Please make sure that you have given us complete contact information including your phone number with country code.</p>
                                        
                                        <p>We look forward to hearing from you!</p>
                
                                        <p>Kind regards,</p>
                                        <p>The ehealthscanner customer support team</p>
                                      </td>
                                    </tr>
                                  </table>
                                </td>
                              </tr>
                
                              <!-- END MAIN CONTENT AREA -->`
            + EmailTemplate.FOOTER +
            `</table>
                <!-- END CENTERED WHITE CONTAINER --></div>
                        </td>
                        <td>&nbsp;</td>
                      </tr>
                    </table>
                  </body>
                </html>
                `
        };
    }

    private _getattachmentsPDF(pathName: string, filename: string): Attachment {
        return MakeAttachmentService().createAttachment(pathName, filename);
    };

    private _getPersonsInfo(persons: any): string {
        let personsInfo = "";

        for (let i = 1; i < persons.length; i++) {
            personsInfo += "Person " + (i + 1) + " age: " + persons[i].age + "\n";
            personsInfo += "Person " + (i + 1) + " gender: " + persons[i].gender + "\n";
            personsInfo += "Person " + (i + 1) + " relationship: " + persons[i].relationShip + "\n";
        }

        return personsInfo;
    }

}
