// helpers/IgnoreOrThrowLogger.ts
/**
 * Created by jose on 27/04/2016.
 */

export type LogLevel = 'debug' | 'info' | 'notice' | 'warn' | 'error' | 'crit' | 'alert' | 'emerg';

export class IgnoreOrThrowLogger {
    public log(level: LogLevel, msg: string): void {
        switch (level) {
            case "debug":
            case "info":
            case "notice":
                break;
            default:
                throw new Error(`${level}: ${msg}`);
        }
    }
    public debug(msg: string): void {
        this.log("debug", msg);
    }
    public info(msg: string): void {
        this.log("info", msg);
    }
    public notice(msg: string): void {
        this.log("notice", msg);
    }
    public warn(msg: string): void {
        this.log("warn", msg);
    }
    public error(msg: string): void {
        this.log("error", msg);
    }
    public crit(msg: string): void {
        this.log("crit", msg);
    }
    public alert(msg: string): void {
        this.log("alert", msg);
    }
    public emerg(msg: string): void {
        this.log("emerg", msg);
    }
}
