// migrations/000000_0_initial_schema.ts
module.exports = {
    up: function (migration: any) {
        migration.schema.createTable(
            "company_plans",
            {
                uuid: {
                    allowNull: false,
                    primaryKey: true,
                    type: migration.Sequelize.STRING(191),
                },
                plan_case_id: {type: migration.Sequelize.STRING(250)},
                plan_title: {type: migration.Sequelize.STRING(250)},
                plan_name: {type: migration.Sequelize.STRING(250)},
                plan_id: {type: migration.Sequelize.STRING(250)},
                company_name: {type: migration.Sequelize.STRING(250)},
                area_of_cover: {type: migration.Sequelize.STRING(1000)},
                master_areas_of_cover: {type: migration.Sequelize.STRING(50)},
                list_of_sub_zone_of_cover_countries_ids: {type: migration.Sequelize.STRING(1000)},
                area_of_residence_plan_availability: {type: migration.Sequelize.STRING(250)},
                list_of_residence_plan_availability_country_ids: {type: migration.Sequelize.STRING(1000)},
                age_minimum: {type: migration.Sequelize.INTEGER(5)},
                age_maximum: {type: migration.Sequelize.INTEGER(5)},
                currency: {type: migration.Sequelize.STRING(10)},
                annual_premium: {type: migration.Sequelize.INTEGER},
                excess: {type: migration.Sequelize.INTEGER},
                maximum_cover: {type: migration.Sequelize.INTEGER},
                maximum_cover_periodicity: {type: migration.Sequelize.STRING(50)},
                excess_periodicity: {type: migration.Sequelize.STRING(50)},
                outpatient_cover_maximum: {type: migration.Sequelize.STRING(50)},
                outpatient_cover_periodicity: {type: migration.Sequelize.STRING(250)},
                outpatient_cover: {type: migration.Sequelize.BOOLEAN},
                outpatient_cover_information: {type: migration.Sequelize.STRING(250)},
                dental_cover_maximum: {type: migration.Sequelize.STRING(50)},
                dental_cover_periodicity: {type: migration.Sequelize.STRING(50)},
                dental_cover: {type: migration.Sequelize.BOOLEAN},
                dental_cover_information: {type: migration.Sequelize.STRING(250)},
                maternity_cover_maximum: {type: migration.Sequelize.STRING(50)},
                maternity_cover_periodicity: {type: migration.Sequelize.STRING(50)},
                maternity_cover: {type: migration.Sequelize.BOOLEAN},
                maternity_cover_information: {type: migration.Sequelize.STRING(250)},
                optical_cover_maximum: {type: migration.Sequelize.STRING(50)},
                optical_cover_periodicity: {type: migration.Sequelize.STRING(50)},
                optical_cover: {type: migration.Sequelize.BOOLEAN},
                optical_cover_information: {type: migration.Sequelize.STRING(250)},
                other_highlights: {type: migration.Sequelize.STRING(1000)},
                only_expat_insurance: {type: migration.Sequelize.BOOLEAN},
                logo_url: {type: migration.Sequelize.STRING(250)},
                benefits_schedule_pdf_url: {type: migration.Sequelize.STRING(250)},
                general_conditions_pdf_url: {type: migration.Sequelize.STRING(250)},
                application_form_pdf_url: {type: migration.Sequelize.STRING(250)},
                co_branding: {type: migration.Sequelize.BOOLEAN},
                co_branding_link: {type: migration.Sequelize.STRING(250)},
                electronic_signature_application_link: {type: migration.Sequelize.STRING(250)},
            }, {
                charset: "utf8mb4",
                collate: "utf8mb4_general_ci",
            }
        );

        migration.schema.createTable(
            "countries",
            {
                country_name: {
                    allowNull: false,
                    primaryKey: true,
                    type: migration.Sequelize.STRING(50),
                },
                country_id: {
                    type: migration.Sequelize.INTEGER(3),
                    unique: true,
                },
            }, {
                charset: "utf8mb4",
                collate: "utf8mb4_general_ci",
            }
        );

        migration.schema.createTable(
            "master_areas_of_cover",
            {
                master_area_name: {
                    allowNull: false,
                    primaryKey: true,
                    type: migration.Sequelize.STRING(191),
                },
                master_area_id: {
                    type: migration.Sequelize.STRING(5),
                    unique: true,
                },
            }, {
                charset: "utf8mb4",
                collate: "utf8mb4_general_ci",
            }
        );
    },
    down: function (migration: any) {
        migration.dropTable("company_plans");
        migration.dropTable("countries");
        migration.dropTable("master_areas_of_cover");
    },
};
