// stubs/domain/RecordHelper.ts
export function MakeFakeRecordInstance() {
    return {
        get: function () {
            //
        },
        where: function () {
            //
        },
        toString: function () {
            //
        },
        setDataValue: function () {
            //
        },
        set: function () {
            //
        },
        setAttributes: function () {
            //
        },
        previous: function () {
            //
        },
        _setInclude: function () {
            //
        },
        save: function () {
            //
        },
        reload: function () {
            //
        },
        validate: function () {
            //
        },
        hookValidate: function () {
            //
        },
        update: function () {
            //
        },
        updateAttributes: function () {
            //
        },
        destroy: function () {
            //
        },
        restore: function () {
            //
        },
        increment: function () {
            //
        },
        decrement: function () {
            //
        },
        equals: function () {
            //
        },
        equalsOneOf: function () {
            //
        },
        setValidators: function () {
            //
        },
        toJSON: function () {
            //
        },
    };
}
