// express/Server.ts
/**
 * Created by kevin on 7/7/15.
 */
const express = require("express");
const winston = require("winston");
import {ExpressMiddleware} from "express/middlewares/ExpressMiddleware";

export default class Server {

    private decorators: ExpressMiddleware[];
    private basicListener: any;
    private logger: any;
    private exp: any;

    constructor(decorators: ExpressMiddleware[],
                basicListener = express(),
                logger = winston,
                exp = express) {
        this.decorators = decorators;
        this.basicListener = basicListener;
        this.logger = logger;
        this.exp = exp;
    }

    get listener() {
        return this.basicListener;
    }

    public listen(port: number) {
        if (undefined === port) {
            throw new Error("You need to specify the port");
        }

        if (this.decorators.length === 0) {
            this.logger.warn("There are not supplied decorators to the server, so it\"s not going to listen any request.");
        }

        try {
            this.decorators.forEach((e: ExpressMiddleware) => {
                return e.setup(this.listener);
            });
        } catch (e) {
            throw e;
        }
        this.listener.listen(port);
        this.logger.info(`SERVER LISTENING ON PORT: ${port}`);
    }
}
